README.md
(→ [Japanese version](http://www.q-eng.imat.eng.osaka-cu.ac.jp/~ippei/html/gomiko/index-ja.html))

# Overview

Gomiko provides a mechanism to temporarily store files rather than delete them, such as the Windows snap box.
[Implement UNIX Technique / Shell](http://www.q-eng.imat.eng.osaka-cu.ac.jp/~ippei/html/unixtodaystips/shell.html#)
As you have done in the past,
I installed the dust box on the shell.
That's 99% of my requirements,
If you have been using it for a long time, you may have felt a shortage as follows.

* I want to hand lightly.
* If you delete a file in the program, you also want to enter the dust box.

Gomiko provides the gomiko command.
The gomiko command has the ability to move files to ~/.trash with changes in rm.
It also displays the history of deletion and undo.
Gomiko as a Gem provides the Gomiko library.
The Ruby program is designed to handle the dust box as well.

## before the name

The name "gomiko" was named as "Japanese box" in Japanese.
If you think that it is "a shrine maiden who plays go", is it cute?

# Install

Rubygems is used.
You can use Rubygems to install it as follows:

    gem install gomiko

## Settings (you like)

The setting is basically unnecessary.
Also generates ~/.trash automatically.
If ~/.trash is present before gomiko and the file is there,
I think that at least a few of them will not work,
It would be more difficult to leave it empty.

If you decide to use gomiko in earnest,
I think you should make rm an alias for gomiko rm.
I have set the following in ~/.zshrc:

    alias rm='gomiko rm'

I think that undo and empty subcommands should be typed like gomiko undo,
I have not created any aliases in particular.


# Command

The gomiko command is available.
Various functions are provided as this subcommand.

## gomiko help

If you do not specify an argument, or if you specify the subcommand help, a helper is displayed.

    % gomiko

    % gomiko help

If you specify a subcommand name after help, as in the following example, the help for the subcommand rm is:

    % gomiko help rm


## gomiko rm

    % gomiko rm foo.txt

It is replaced with the normal rm command.
To change the deletion
Move to the ~/.trash directory.
The moved file name is
"~/.trash/date-time/original full path"
It is becoming more and more common.
This
"Date-time" is the timing when gomiko was started,
It is not the time when the movement was performed internally.
For example, if `gomiko rm *` "deletes a large number of files and takes more than a second,
It can be entered in one "date-time".

You can specify more than one file at a time, as follows:

    % gomiko rm foo.txt bar.txt */*.txt

You can also move the directory to the dust box without options, as shown below.

    % gomiko rm dir1/ dir2/

When multiple gomiko rms are performed at the same time,
The first is
It will be the directory name `` ~/.trash/date - time ``,
From now on,
It is created in serial number after time like "~/.trash/date-time-1".
It is a mechanism that is similar to the Locked Directory of Exclusive Control,
Because only the process that generated the directory is in the directory,
It is also safe for file systems that share NFS over multiple machines.

After that, the file directory under ~/.trash/ will be called "trash ID directory"
The directory name is called the "ID".
Even if it was created manually rather than via gomiko rm,
It can be used as a trash ID directory, trash ID.
(Whether undo or other commands are intended to work is another matter.)


## gomiko undo

The subcommand undo provides the un-delete function for file deletion.
If you run it without arguments as in the following, you will undo the most recent contents of the trash ID directory.

    % gomiko undo

Because undo removes the appropriate discovery ID directory,
If you run gomiko undo, you will then unhide the new Discovery ID directory.
In other words, running gomiko undo will continue to remove the deletion.

If the file exists in the original path that you deleted,
It does not issue a warning about the file in question and does not perform the move.
Other files are returned.

    % gomiko undo
    Normal file already exist: /home /ippei /tmp /gomiko /0
    Can not undo: /home/ippei/.trash/20170623-133644

You can specify a false ID to execute undo on the argument.

    % gomiko undo 20170623-133644

(In this case, the expansion of the memory card can not be normal.)

    % gomiko undo 20170623-133644 20170623-123456

You can also specify the path

    % gomiko undo ~/.trash/20170623-*

## gomiko empty

Empty the contents of the ~/.trash directory.
If the argument is omitted, all file directories in ~/.trash are considered.
Gomiko empty is also a file directory that is in ~/.trash without gomiko rm
It is targeted.

You can specify a false ID in the argument.

    % gomiko empty 20170628-112425-2

It can be specified with a path instead of a false ID.

    % gomiko empty ~/.trash/20170628-112425-2

### --mtime option

It is recommended that you periodically delete the contents of ~/.trash.
At that point, if you empty the file at some stage, the file you moved to ~/.trash
It is also possible to delete it.
If you look at the timestamp, for example, delete the file after seven days,
It is a routine method.
The --mtime option does this.

    % gomiko empty --mtime=-7

, Delete the item that has passed 7 days or more.
Because we decided on the notation using the -mtime option of find,
Normally you will specify a negative number.
The default value of the - mtime option argument is 0,
All previously created trash ID directories are considered.
This time stan- dard,
We are looking at the time stamp (update time) of the geo ID directory.
Because you did not get the date from the file name,
Changing the file name does not change the date judgment.

If you use the --mtime option with the argument id of the argument,
Select an AND condition that matches both.
(Note: Even if it is considered to be an OR condition,
 If you do OR, you only need to issue the command twice,
 Because the AND condition is between the hand and the hand, the value is still.
 I do not think that it should be used together.)

### --quiet option

You will not need to log the standard output when running with cron.
In such cases, the --quiet option can suppress the standard output.

    % gomiko empty --quiet

When you run the empty subcommand manually
It is more pleasant to see something, and by default, --quiet is off.

## gomiko ls

Displays the contents of ~/.trash.

    % gomiko ls
    size date-time-id    typical-path[ ...]
    20K  20170623-125159 /home/ippei/tmp/gomiko/0
    20K  20170623-125810 /home/ippei/tmp/gomiko/5 ...
    84K  20170627-084500 /home/ippei/git/gomiko/test/gomiko/tmp/
    24K  20170627-091649 /home/ippei/git/gomiko/a/
    32K  20170627-091712 /home/ippei/git/gomiko/test/gomiko/b (exist in original path)
    100K 20170627-092407 /home/ippei/git/gomiko/test/gomiko/tmp/

The first line is the title line of information.
The second and third lines are data.
The size of the first column is the size that can be acquired by du.
The second column is representative of the path that is assumed to be deleted.
This inference is based on the recursive search of the trash ID directory,
And it presents a file directory that does not exist on the actual full path.
Following the second column, "..." means that there is an assumed path other than the representative,
In other words, it is assumed that you have moved multiple files into a single box with a single command.

If the original path that you deleted contains a file that already exists,
It outputs (may exist newer file) as follows.

    20K  20170623-125810 /home/ippei/tmp/gomiko/5 ...(may exist newer file)

You can specify a false ID for the argument.

    % gomiko ls 20170623-125159 20170623-125810
    size date-time-id    typical-path[ ...]
    20K  20170623-125159 /home/ippei/tmp/gomiko/0
    20K  20170623-125810 /home/ippei/tmp/gomiko/5 ...


The path is OK even if it is a bad ID.

    % gomiko ls ~/.trash/20170623-125*
    size date-time-id    typical-path[ ...]
    20K  20170623-125159 /home/ippei/tmp/gomiko/0
    20K  20170623-125810 /home/ippei/tmp/gomiko/5 ...

### --long, -l option

The --long (-l) option prints more information.

    % gomiko ls --long
    ------------------------------------------------------------
    size      : 20K
    id        : 20170628-112425
    guess path: /home/ippei/tmp/gomiko/0
    +----- filetype in trash
    | +--- filetype in original path
    | | +- original path
    / / /home
    / / /home/ippei
    / / /home/ippei/tmp
    / / /home/ippei/tmp/gomiko
    .   /home/ippei/tmp/gomiko/0

    ------------------------------------------------------------
    size      : 20K
    id        : 20170628-112425-1
    guess path: /home/ippei/tmp/gomiko/1
    +----- filetype in trash
    | +--- filetype in original path
    | | +- original path
    / / /home
    / / /home/ippei
    / / /home/ippei/tmp
    / / /home/ippei/tmp/gomiko
    .   /home/ippei/tmp/gomiko/1

    ------------------------------------------------------------
    size      : 20K
    id        : 20170628-112425-3
    guess path: /home/ippei/tmp/gomiko/3
    +----- filetype in trash
    | +--- filetype in original path
    | | +- original path
    / / /home
    / / /home/ippei
    / / /home/ippei/tmp
    / / /home/ippei/tmp/gomiko
    .   /home/ippei/tmp/gomiko/3

It is divided by a horizontal line for each gummy ID.
Capacity (size),
The ID (id)
The guessed path (guess path)
Below is the information for all the directory files stored in the trash ID directory,
The first element of each line is the file type in the geo ID directory,
The next element is the file type of the original path on the current file tree.
The meaning of the letter that indicates the file type is as follows.

* "/" Directory
* "." Normal file
* " "(Blank) indicates that the file does not exist


# Library

Gomiko.new generates a Gomiko object,
By using Gomiko#throw,
You can use the dust box from the program.


# Rejected features

## It takes a long time to move to a local folder when deleting an NFS share file

Solutions to this need are difficult.
The problem is how to place the dust box directory equivalent to ~/.trash.
The df command provides information about the file system,
It is difficult to determine automatically due to various factors such as sym- bolic links and permutations.
For example, if you followed the expanded path of a symlink,
To the top-level directory that has the same file system or permission to write to itself
Let's imagine a policy of creating a gummy box directory.
If you delete it in /tmp, you will create a directory called /tmp/.trash.
I can not say that it is an amiable good.
In addition, there are multiple dust box directories,
It becomes difficult to know where the target file is stored.
The benefits are far less complex and complex, and there are fewer demerits.
It is possible to create an option to specify the location of the glove box directory,
There is no plan to make such an option at this time.


## mkdir and directory specification

Every time you run gomiko (or whenever you throw Gomiko#throw internally), a new gomi ID directory is created,
The granularity of gomiko undo is in that unit.
When you delete a file from a program,
There is a request that "we want to go to a common false ID directory in one process".
To solve this problem, create a gomi ID directory with gomiko mkdir,
You can specify gomiko rm or Gomiko#throw with that directory.
Although technological difficulties are not difficult,
This will cause problems as a program's log- ics.
The problem is that multiple file deletions can be the same path in a single trash ID directory.
Assuming that you have implemented the - id option in gomiko mkdir and gomiko rm,
Let's consider it in a shell script.

    DIR=`gomiko mkdir`#Creates a new ID directory and returns its ID
    touch a
    gomiko rm --id=$ DIR a # Specify the id ID directory and drag it there.
    touch a
    gomiko rm --id=$ DIR a # It should be loaded into the same path of the same god ID directory.

The second gomiko rm causes duplication.
Because the file trees at different points in time are grouped together,
Such problems may arise.
As a program, you should position the geo-ID directory as a snapshot of the file tree at the time of the deletion.

From these, there is no plan to add this function.

## redo function

Currently, gomiko uses the data in ~/.trash files as its data
Processing is being performed.
By not having the history as a separate file,
And preserves the uniqueness of the information.
To do a redo, you need to remove the undo file from ~/.trash,
It is necessary to place a record somewhere.
As a result, the place of information becomes less centralized,
You have to be careful about maintaining consistency.
Even if gomiko itself does not have the redo function,
The same can be done by using the shell's histories.
Because I think that it is a function with a lot of effort in terms of work,
There is no plan to add this function.

## Using the Directory Creation Time

And the directory in the trash ID directory.
If we can compare the generation time of the source directories,
It will be possible to deduce more accurate delete targets.
For example, if you have a file called /a /b /c.txt,
Suppose you deleted it with /a /b.
The ending file directory is
/A and ~/.trash/ID/a/b.
If you then generate /a/b,
The /a/b should be later than the creation time by ~/.trash /false ID /a /b.
On the other hand, if you look at a, the ~/.trash/false ID /a generated by deletion should be later than the /a that was already present at the time of deletion.
You should be able to determine whether this is the directory generated after the deletion,
On Linux mainstream ext4 file systems
The file directory creation time is not recorded in the timestamp.
Therefore, this can not be used.
We will also consider when the file system that can be used at creation time is mainstream.


# Contributing to gomiko
 
* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet.
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it.
* Fork the project.
* Start a feature/bugfix branch.
* Commit and push until you are happy with your contribution.
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the Rakefile, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

# Copyright

Copyright (c) 2017 ippei94da. See LICENSE.txt for
further details.



CHANGES
= Gomiko changelog

== Version 0.0.2 [2018-02-20] released

* small bugfix

== Version 0.0.1 [2017-06-29] released

* Bugfix a problem: enable to remove deadlink.
* Bugfix a problem that 'gomiko ls' not show symlink deletion.
* Avoid exception when 'gomiko rm .'

== Version 0.0.0 [2017-06-29] released

* initial release



test/helper.rb
require 'simplecov'

module SimpleCov::Configuration
  def clean_filters
    @filters = []
  end
end

SimpleCov.configure do
  clean_filters
  load_adapter 'test_frameworks'
end

ENV["COVERAGE"] && SimpleCov.start do
  add_filter "/.rvm/"
end
require 'rubygems'
require 'bundler'
begin
  Bundler.setup(:default, :development)
rescue Bundler::BundlerError => e
  $stderr.puts e.message
  $stderr.puts "Run `bundle install` to install missing gems"
  exit e.status_code
end
require 'test/unit'

$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
require 'gomiko'

class Test::Unit::TestCase
end


test/test_gomibako.rb
#! /usr/bin/env ruby
# coding: utf-8

require "pp"
require "helper"

class Gomibako
  public :graft
end

class TC_Gomibako < Test::Unit::TestCase

  TRASHDIR = 'test/gomibako_tmp/trash'
  #$stdout = StringIO.new

  def setup
    #FileUtils.rm_rf 'test/gomibako_tmp/trash'
    FileUtils.rm_rf 'test/gomibako_tmp'
    @g00 = Gomibako.new(dir: TRASHDIR, verbose: false)
  end

  def teardown
    FileUtils.rm_rf 'test/gomibako_tmp'
  end

  def test_initialize
    FileUtils.rm_rf 'test/gomibako_tmp/trash'
    assert_false(FileTest.directory? TRASHDIR)
    Gomibako.new(dir: TRASHDIR, verbose: false)
    assert(FileTest.directory? TRASHDIR)
  end

  def test_throw
    # remove file
    a_relpath = 'test/gomibako_tmp/a.txt'
    a_fullpath = File.expand_path 'test/gomibako_tmp/a.txt'
    a_fullpath_dirname = File.dirname a_fullpath
    #File.open(a_relpath, 'w')
    FileUtils.touch a_relpath
    @g00.throw(paths: [a_relpath], time: Time.new(2017, 1, 23, 12, 34, 56), verbose: false)
    assert(FileTest.directory? "#{TRASHDIR}/20170123-123456#{a_fullpath_dirname}")
    assert_false(FileTest.exist? a_relpath)
    assert(FileTest.exist? ("#{TRASHDIR}/20170123-123456#{a_fullpath_dirname}"))

    # remove directory
    a_relpath = 'test/gomibako_tmp/b/a.txt'
    a_fullpath = File.expand_path 'test/gomibako_tmp/b/a.txt'
    a_fullpath_dirname = File.dirname a_fullpath
    FileUtils.mkdir_p 'test/gomibako_tmp/b'
    FileUtils.touch a_relpath
    @g00.throw(paths: [a_relpath], time: Time.new(2017, 1, 23, 12, 34, 56), verbose: false)
    assert(FileTest.directory? "#{TRASHDIR}/20170123-123456#{a_fullpath_dirname}")
    assert_false(FileTest.exist? a_relpath)
    assert(FileTest.exist? ("#{TRASHDIR}/20170123-123456#{a_fullpath_dirname}"))

    # include not exist file
    #io = StringIO.new
    @g00.throw(paths: [a_relpath, 'not_exist'],
               time: Time.new(2017, 1, 23, 12, 34, 56),
               verbose: false)
    assert(FileTest.exist? (a_relpath))
  end
  
  def test_empty
    a_relpath = 'test/gomibako_tmp/a.txt'
    a_fullpath = File.expand_path 'test/gomibako_tmp/a.txt'
    #File.open(a_relpath, 'w')
    FileUtils.touch a_relpath
    @g00.throw(paths: [a_relpath], time: Time.new(2017, 1, 23, 12, 34, 56), verbose: false)

    assert(File.exist? "#{TRASHDIR}/20170123-123456#{a_fullpath}")
    @g00.empty(verbose: false)
    assert_false(File.exist? "#{TRASHDIR}/20170123-123456#{a_fullpath}")
  end

  def test_empty_before
    # before option
    a_relpath = 'test/gomibako_tmp/a.txt'
    a_fullpath = File.expand_path 'test/gomibako_tmp/a.txt'
    #File.open(a_relpath, 'w')
    FileUtils.touch a_relpath
    @g00.throw(paths: [a_relpath],
               time: Time.new(2017, 1, 23, 12, 34, 56),
               verbose: false)

    assert(File.exist? "#{TRASHDIR}/20170123-123456#{a_fullpath}")
    @g00.empty(before: 5,
               time: Time.new(2017, 1, 27, 12, 34, 56),
               verbose: false)
    assert(File.exist? "#{TRASHDIR}/20170123-123456#{a_fullpath}")
    @g00.empty(before: 3,
               time: Time.new(2017, 1, 27, 12, 34, 56),
               verbose: false)
    assert(File.exist? "#{TRASHDIR}/20170123-123456#{a_fullpath}")
  end

  #def test_undo
  #  FileUtils.rm_rf "test/gomibako_tmp/undo"
  #  #a_relpath = 'a/b/c.txt'
  #  #b_relpath = 'a/b/d.txt'
  #  #b_fullpath = File.expand_path 'test/gomibako_tmp/b.txt'
  #  FileUtils.touch 'a/b/c.txt'
  #  FileUtils.touch 'a/b/d.txt'
  #  @g00.throw(paths: ['a/b/c.txt'], time: Time.new(2017, 1, 23, 12, 34, 56), verbose: false)
  #  @g00.throw(paths: ['a/b/d.txt'], time: Time.new(2017, 1, 23, 12, 34, 57), verbose: false)

  #  assert(File.exist? "#{TRASHDIR}/20170123-123456/a/b/c.txt")
  #  assert(File.exist? "#{TRASHDIR}/20170123-123457/a/b/d.txt")
  #  @g00.undo(verbose: false, dst_root: "test/gomibako_tmp/undo")
  #  assert      (File.exist? "#{TRASHDIR}/20170123-123457/a/b/c.txt")
  #  assert_false(File.exist? "#{TRASHDIR}/20170123-123457/a/b/d.txt")
  #  assert      (File.exist? "test/gomibako_tmp/undo/a/b/d.txt")
  #end

  def test_graft1
    FileUtils.rm_rf(  'test/gomibako_tmp/graft')
    FileUtils.mkdir_p('test/gomibako_tmp/graft/src/a/b/c')
    FileUtils.touch(  'test/gomibako_tmp/graft/src/a/b/c/d.txt')
    @g00.graft('test/gomibako_tmp/graft/src',
               '',
               dst_root: "test/gomibako_tmp/graft/dst",
              verbose: false)

    assert(FileTest.directory?('test/gomibako_tmp/graft/dst/a/b/c/'))
    assert(FileTest.file?('test/gomibako_tmp/graft/dst/a/b/c/d.txt'))
  end

  def test_graft2
    FileUtils.rm_rf(  'test/gomibako_tmp/graft')
    FileUtils.mkdir_p('test/gomibako_tmp/graft/src/a/b/c')
    FileUtils.mkdir_p('test/gomibako_tmp/graft/dst/a')
    FileUtils.touch(  'test/gomibako_tmp/graft/src/a/b/c/d.txt')
    #
    @g00.graft('test/gomibako_tmp/graft/src',
               'a',
               dst_root: "test/gomibako_tmp/graft/dst",
              verbose: false)
    assert(FileTest.directory?('test/gomibako_tmp/graft/dst/a/b/c/'))
    assert(FileTest.file?('test/gomibako_tmp/graft/dst/a/b/c/d.txt'))
  end

  def test_graft3
    FileUtils.rm_rf(  'test/gomibako_tmp/graft')
    FileUtils.mkdir_p('test/gomibako_tmp/graft/src/a/b/c')
    FileUtils.mkdir_p('test/gomibako_tmp/graft/dst/a/b/c')
    FileUtils.touch(  'test/gomibako_tmp/graft/src/a/b/c/d.txt')
    #
    @g00.graft('test/gomibako_tmp/graft/src',
               'a',
               dst_root: "test/gomibako_tmp/graft/dst",
              verbose: false)
    assert(FileTest.directory?('test/gomibako_tmp/graft/dst/a/b/c/'))
    assert(FileTest.file?('test/gomibako_tmp/graft/dst/a/b/c/d.txt'))
  end
end



lib/gomibako.rb
#! /usr/bin/env ruby
# coding: utf-8

require 'fileutils'
require 'pathname'
require 'tefil'
require 'find'

#
#
#
class Gomibako

  def initialize(dir: nil, verbose: true)
    if dir
      @trashdir = dir
    else
      if ENV['UID'] == 0 # this is needed for 'su -m'
        @trashdir ='/root/.trash' 
      else
        @trashdir = ENV['HOME'] + "/.trash"
      end
    end
    FileUtils.mkdir_p(@trashdir)
  end

  def throw(paths: , time: Time.new, verbose: true)
    paths.each do |path|
      unless FileTest.exist? path
        puts "gomibako rm: cannot remove '#{path}': No such file or directory" if verbose
        exit
      end
    end

    trash_subdir = @trashdir + time.strftime('/%Y%m%d-%H%M%S')
    FileUtils.mkdir_p(trash_subdir)
    paths.each do |path|
      dst = trash_subdir + File.expand_path(path)
      dst_dir = File.dirname dst
      FileUtils.mkdir_p(dst_dir)
      FileUtils.mv(path, dst_dir + '/', :verbose => verbose)
    end
  end

  def empty(before: 0, time: Time.now, verbose: true)
    Dir.glob("#{@trashdir}/*").each do |path|
      if time -  File.mtime(path) > 86400 * before
        FileUtils.rm_rf(path, :verbose => verbose)
      end
    end
  end

  def undo(verbose: true, dst_root: '/')
    if Dir.glob("#{@trashdir}/*").empty?
      puts "Nothing to undo in #{@trashdir}"
      exit
    end
    dst_dir = Dir.glob("#{@trashdir}/*").sort_by { |path| File.ctime path }[-1]

    begin
      Dir.glob("#{dst_dir}/*").sort.each do |path|
        #rsync might be an idea.
        #but it has a problem that it cannot merge to thepath without permission.
        graft(dst_dir, '', dst_root: dst_root)
      end
      FileUtils.rm_rf dst_dir # risky?
    rescue
      puts "Cannot undo: #{dst_dir}"
    end
  end

  def ls
    results = [['size', '[date-time dir]/path[ ...]']]
    Dir.glob("#{@trashdir}/*").sort.each do |path|
      tmp = []
      tmp << `du --human-readable --max-depth=0 #{path}`.split(' ')[0] # size

      not_exist_files = Find.find(path).select do |subpath|
        fullpath = subpath.sub(/^#{path}/, '')
        ! FileTest.exist?(fullpath)
      end
      not_exist_files.shift # path for root dir
      not_exist_files.map! {|v| v.sub(/^#{@trashdir}\//, '')}
      tmp << not_exist_files[0]
      tmp[-1] += ' ...' if not_exist_files.size > 2

      results << tmp
    end
    if results.size > 1
      Tefil::ColumnFormer.new.form(results)
    else
      puts "Nothing in ~/.trash"
    end
  end

  private

  # e.g., root_path = ~/.trash/20170123-012345
  #       path      = home/ippei/foo
  def graft(src_root, path, dst_root: '/', verbose: true)
    src_root = Pathname.new(src_root)
    path     = Pathname.new(path)
    dst_path = Pathname.new(dst_root) + path

    if FileTest.directory? (dst_path)
      Dir.glob("#{src_root + path}/*") do |next_path|
        next_path = Pathname.new(next_path).relative_path_from(src_root)
        graft(src_root, next_path, dst_root: dst_root, verbose: verbose)
      end
    elsif FileTest.exist? (dst_path)
      puts "normal file already exist: #{dst_path}" if verbose
    else
      FileUtils.mv(src_root + path, dst_path, noop: false, verbose: verbose )
    end
    return
  end

end



lib/gomiko.rb
#! /usr/bin/env ruby
# coding: utf-8

require 'fileutils'
require 'pathname'
require 'tefil'
require 'find'
require 'yaml'

#
#
#
class Gomiko

  attr_reader :trashdir

  #class NotFoundError < StandardError; end 

  def initialize(dir: nil, verbose: true)
    if dir
      @trashdir = dir
    else
      if ENV['UID'] == 0 # this is needed for 'su -m'
        @trashdir ='/root/.trash' 
      else
        @trashdir = ENV['HOME'] + "/.trash"
      end
    end
    FileUtils.mkdir_p(@trashdir)
  end

  # If paths includes exist and not exist files,
  # throw all exist file and report not exist files.
  def throw(paths: , time: Time.new, verbose: true, io: $stdout)
    paths = paths.select do |path|
      flag = FileTest.symlink?(path) || FileTest.exist?(path) # for deadlink
      unless flag
        if verbose
          io.puts "gomiko rm: cannot remove '#{path}': No such file or directory"
        end
      end
      flag
    end
    return if paths.empty?

    trash_subdir = mkdir_time(time)
    File.open(trash_subdir + ".yaml", "w") do |yaml_io|
      YAML.dump( { 'pwd' => ENV["PWD"], 'paths' => paths }, yaml_io)
    end

    paths.each do |path|
      dst = trash_subdir + File.expand_path(path)
      dst_dir = File.dirname dst
      FileUtils.mkdir_p(dst_dir)
      if path == '.'
        io.puts "gomiko rm: failed to remove '.': Invalid argument" if verbose
        next
      else
        FileUtils.mv(path, dst_dir + '/')
        io.puts "mv #{path} #{dst_dir}" if verbose
        File.utime(time, time, trash_subdir)
      end
    end
  end

  def empty(ids: [], mtime: 0, time: Time.now, verbose: true, io: $stdout)
    if ids.empty?
      tgts = (dir_list + yaml_list.map{|v| v.sub(".yaml", "")}).sort.uniq
    else
      tgts = ids.map {|v|  path2id(v.sub(".yaml", ""))}
    end
    tgts = tgts.select { |v|
      begin
        str2datetime(v) - time < 86400 * mtime
      rescue Errno::ENOENT
        io.puts "Not found: #{v}"
      end
    }
    tgts.map! {|v|  @trashdir + '/' + v}
    if tgts.empty?
      io.puts "No directory was emptied."
    else
      tgts.each do |path|
        ["", ".yaml"].each do |ext|
          new_path = path + ext
          FileUtils.rm_rf(new_path, :verbose => verbose) if File.exist? new_path
        end
      end
    end
  end

  def undo(id, verbose: true, io: $stdout)
    id = path2id id
    fullpath = Pathname.new(@trashdir) + id
    Dir.glob("#{fullpath}/*").sort.each do |path|
      graft(fullpath, '', dst_root: '/', verbose: verbose)
    end
    if Dir.glob("#{fullpath}/**/*").find {|path| FileTest.file? path}
      io.puts "Unable to complete undo: #{fullpath}"
    else
      FileUtils.rm_rf fullpath # risky?
    end
  end

  # Example of return data:
  # [
  #   236K,
  #   20170623-021233,
  #   [/home/ippei/tmp/a.txt, /home/ippei/tmp/b.txt ]
  # ]
  def info(id)
    id = path2id id
    cur_trash_dir = Pathname.new(@trashdir) + id
    results = []
    results << `du --human-readable --max-depth=0 #{cur_trash_dir}`.split(' ')[0]
    results << id

    trash_paths = Dir.glob("#{cur_trash_dir}/**/*", File::FNM_DOTMATCH).sort
    trash_paths = trash_paths.select { |t_path| ! /\/\.$/.match t_path } # '/.', で終わるのを除外

    candidates = [] # fo rm target
    results_long = []
    additions = []
    #flag_conflict = false
    flag_include_file = false
    trash_paths.each do |trash_path|
      orig_path = trash_path.sub(/^#{cur_trash_dir}/, '')
      trash_type = ftype_str(trash_path)
      flag_include_file = true unless File.directory? trash_path
      if FileTest.exist? orig_path
        orig_type = ftype_str(orig_path)
        if File.ftype(trash_path) != File.ftype(orig_path)
          #flag_conflict = true
          additions << 'conflict'
          candidates << trash_path
        end
      else
        candidates << trash_path
        orig_type   = ' '
      end
      results_long << [ trash_type, orig_type,
                        trash_path.sub(/^#{cur_trash_dir}/, '') ]
    end
    unless flag_include_file
      additions << 'only directory'
      candidates << trash_paths[-1]
    end

    results <<  YAML.load_file(cur_trash_dir.to_s + ".yaml")['paths'][0]
    ## if no candidate, last file is adopted.
    if trash_paths.empty?
      #results << '(empty)'
      results << []
    else
      additions << 'conflict' if candidates.empty?
      #candidates = candidates.map    {|path|
      #  tmp = path.sub(/^#{cur_trash_dir}/, '')
      #  tmp += '/' if FileTest.directory? path
      #  tmp
      #}
      #results << candidates[0]

      ## output '...' when multiple.
      candidates = candidates.select{|pa| ! pa.include? candidates[0]}
      results[-1] += ' ...' unless candidates.empty?
      #results[2] += ' (conflict)' if flag_conflict
      results[2] += ' (' + additions.join(',') + ')' unless additions.empty?
      results << results_long
    end
    results
  end

  # ls, yaml_list
  # return stocked yaml.
  def yaml_list
    Dir.glob("#{@trashdir}/*.yaml").map { |path|
      path.sub(/^#{@trashdir}\//, '').split[0]
    } . sort
  end

  # ls, dir_list
  # return stocked directories.
  def dir_list
    Dir.glob("#{@trashdir}/*").select { |path|
      FileTest.directory? path
    } .map { |path|
      path.sub(/^#{@trashdir}\//, '').split[0]
    } . sort
  end

  private

  def str2datetime(str)
    /(\d{4})(\d{2})(\d{2})-(\d{2})(\d{2})(\d{2})/ =~ str
    Time.new( $1.to_i, $2.to_i, $3.to_i, $4.to_i, $5.to_i, $6.to_i)
  end

  # e.g., root_path = ~/.trash/20170123-012345
  #       path      = home/ippei/foo
  def graft(src_root, path, dst_root: '/', verbose: true)
    src_root = Pathname.new(src_root)
    path     = Pathname.new(path)
    dst_path = Pathname.new(dst_root) + path

    if FileTest.directory? (dst_path)
      Dir.glob("#{src_root + path}/*") do |next_path|
        next_path = Pathname.new(next_path).relative_path_from(src_root)
        graft(src_root, next_path, dst_root: dst_root, verbose: verbose)
      end
    elsif FileTest.exist? (dst_path)
      puts "normal file already exist: #{dst_path}" if verbose
    else
      FileUtils.mv(src_root + path, dst_path, noop: false, verbose: verbose )
    end
    return
  end

  def mkdir_time(time)
    time_str = time.strftime('/%Y%m%d-%H%M%S')
    dirname = nil

    i = 0
    while ! dirname

      begin
        #self.ls
        try_name = @trashdir + "#{time_str}"
        try_name += "-#{i}" if 1 <= i
        FileUtils.mkdir(try_name)
        dirname = try_name
      rescue Errno::EEXIST
        i += 1
        next
      end
    end
    dirname
  end

  def ftype_str(path)
    if File.directory? path
      result = '/'
    elsif File.symlink? path
      result = '@'
    elsif File.file? path
      result = '.'
    else
      result = '?'
    end
    result
  end

  def path2id(path)
    path.sub(/^#{@trashdir}\//, '')
  end

end

